system/Database/BaseBuilder.php

```diff
@@ -1588,38 +1588,49 @@ public function insertBatch(?array $set = null, ?bool $escape = null, int $batch
 
                 return false; // @codeCoverageIgnore
             }
-        } else {
-            if (empty($set)) {
-                if (CI_DEBUG) {
-                    throw new DatabaseException('insertBatch() called with no data');
-                }
-
-                return false; // @codeCoverageIgnore
+        } elseif (empty($set)) {
+            if (CI_DEBUG) {
+                throw new DatabaseException('insertBatch() called with no data');
             }
 
-            $this->setInsertBatch($set, '', $escape);
+            return false; // @codeCoverageIgnore
         }
 
+        $hasQBSet = ($set === null);
+
         $table = $this->QBFrom[0];
 
         $affectedRows = 0;
+        $savedSQL     = [];
 
-        for ($i = 0, $total = count($this->QBSet); $i < $total; $i += $batchSize) {
-            $sql = $this->_insertBatch($this->db->protectIdentifiers($table, true, null, false), $this->QBKeys, array_slice($this->QBSet, $i, $batchSize));
+        if ($hasQBSet) {
+            $set = $this->QBSet;
+        }
+
+        for ($i = 0, $total = count($set); $i < $total; $i += $batchSize) {
+            if ($hasQBSet) {
+                $QBSet = array_slice($this->QBSet, $i, $batchSize);
+            } else {
+                $this->setInsertBatch(array_slice($set, $i, $batchSize), '', $escape);
+                $QBSet = $this->QBSet;
+            }
+            $sql = $this->_insertBatch($this->db->protectIdentifiers($table, true, null, false), $this->QBKeys, $QBSet);
 
             if ($this->testMode) {
-                $affectedRows++;
+                $savedSQL[] = $sql;
```

Comment:
Some tests may start failing because of that. It will be an easy fix for any developer, but we have to mention this in the changelog and upgrade notes.

Reply:
Added.

Reply:
As I wrote the commit message https://github.com/codeigniter4/CodeIgniter4/pull/5202/commits/0f3088c2786c5bc66f39b1857659ca99c47bcdd0 to make the return type of insertBatch() the same as updateBatch() .

> Now, the result for this method will be quite different from the result we get without the testMode.

I did not intend to change the behavior without the testMode. Am I something missing?